### PR TITLE
Adding the final keyword for EntityManager to enforce no extension

### PR DESCRIPTION
**Pull request by usmanzafar**
Added the final keyword for EntityManager to indicate that extending EntityManager is not a valid extension point as recommended in https://github.com/doctrine/doctrine2/pull/6595 by @Ocramius / @Majkl578 
Let me know what do you guys think, very excited - first time contributing to a project that i use almost everyday for the last 3 years.

### CODE REVIEW

**Comment by Majkl578** [comment]
Unfortunately it won't be _that_ easy as just uncommenting the modifier. There is at least one case that needs to be adressed first: Doctrine\Tests\Mocks\EntityManagerMock currently extends EntityManager. :/

**Comment by Ocramius** [comment]
To be added:

 * [ ] document breaking change in `UPGRADE.md` or equivalent
 * [ ] change test suite so the mock wraps a real `EntityManager` and implements the `EntityManagerInterface`
 * [ ] (maybe) consider killing `EntityManagerDecorator`?

**Comment by usmanzafar** [comment]
Yes, should have checked thats my bad. Let me have a look at this and get back asap.

File: tests/Doctrine/Tests/Mocks/EntityManagerMock.php [hunk 1]
```diff
@@ -23,12 +33,25 @@ class EntityManagerMock extends EntityManager
...
+        $this->em = $em;
+    }
+
     /**
      * {@inheritdoc}
      */
     public function getUnitOfWork()
     {
-        return isset($this->uowMock) ? $this->uowMock : parent::getUnitOfWork();
+        return isset($this->uowMock) ? $this->uowMock : $this->em->getUnitOfWork();
```
Comment (alcaeus) [inline tests/Doctrine/Tests/Mocks/EntityManagerMock.php]: `$this->uowMock ?? $this->em->getUnitOfWork();`

File: tests/Doctrine/Tests/Mocks/EntityManagerMock.php [hunk 2]
```diff
@@ -60,7 +83,7 @@ public function setProxyFactory($proxyFactory)
      */
     public function getProxyFactory()
     {
-        return isset($this->proxyFactoryMock) ? $this->proxyFactoryMock : parent::getProxyFactory();
+        return isset($this->proxyFactoryMock) ? $this->proxyFactoryMock : $this->em->getProxyFactory();
```
Comment (alcaeus) [inline tests/Doctrine/Tests/Mocks/EntityManagerMock.php]: `$this->proxyFactoryMock ?? $this->em->getProxyFactory()`

Comment (usmanzafar) [inline tests/Doctrine/Tests/Mocks/EntityManagerMock.php]: ahh yes! PHP 7 love. Will do it now. thanks @alcaeus :)

**Comment by usmanzafar** [comment]
Hey @alcaeus i have updated the Mock / tests as per the recommendation, i have also merged https://github.com/doctrine/doctrine2/pull/6599 by @Majkl578 as this was necessary for the tests to pass. 
Let me know your thoughts.  Thanks

**Comment by Ocramius** [comment]
Please rebase: the interface patch was merged to `develop`

**Comment by usmanzafar** [comment]
@Ocramius  - done captain!

File: UPGRADE.md [hunk 3]
```diff
@@ -6,6 +6,9 @@
 
 ## BC Break: Removed `EntityManager#flush($entity)` and `EntityManager#flush($entities)`
 
+Final keyword has been added to the ``EntityManager::class`` in order to ensure that EntityManager is not used as valid extension point. Valid extension point should be EntityManagerInterface.
+## BC Break: Removed `EntityManager#flush($entity)` and `EntityManager#flush($entities)`
```
Comment (Ocramius) [inline UPGRADE.md]: Something went wrong in the rebase here, and your BC break heading is gone

File: tests/Doctrine/Tests/ORM/Hydration/ObjectHydratorTest.php [hunk 4]
```diff
@@ -1005,6 +1005,7 @@ public function testScalarQueryWithoutResultVariables($userEntityKey)
      */
     public function testCreatesProxyForLazyLoadingWithForeignKeys()
     {
+        $this->markTestSkipped('Fix with StaticProxy Factory as ProxyFactory::class does not exist anymore');
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Hydration/ObjectHydratorTest.php]: Skipping this is unrelated to your patch

File: tests/Doctrine/Tests/ORM/Hydration/ObjectHydratorTest.php [hunk 5]
```diff
@@ -1054,6 +1055,7 @@ public function testCreatesProxyForLazyLoadingWithForeignKeys()
      */
     public function testCreatesProxyForLazyLoadingWithForeignKeysWithAliasedProductEntity()
     {
+        $this->markTestSkipped('Fix with StaticProxy Factory as ProxyFactory::class does not exist anymore');
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Hydration/ObjectHydratorTest.php]: Same

File: tests/Doctrine/Tests/ORM/QueryBuilderTest.php [hunk 6]
```diff
@@ -792,7 +792,7 @@ public function testMultipleIsolatedQueryConstruction()
     public function testGetEntityManager()
     {
         $qb = $this->em->createQueryBuilder();
-        self::assertEquals($this->em, $qb->getEntityManager());
+        self::assertEquals($this->em->getEM(), $qb->getEntityManager());
     }
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/QueryBuilderTest.php]: Correct the type of the `em` property in this test

File: tests/Doctrine/Tests/ORM/EntityManagerTest.php [hunk 7]
```diff
@@ -222,7 +222,7 @@ public function testTransactionalAcceptsVariousCallables()
 
     public function transactionalCallback($em)
     {
-        self::assertSame($this->em, $em);
+        self::assertSame($this->em->getEM(), $em);
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/EntityManagerTest.php]: Correct the type of the `em` property in this test

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by usmanzafar** [comment]
Hey @Ocramius apologies for not replying earlier, i have fixed the tests and updated others as per your advice. Let me know your thoughts on this. Thanks

File: lib/Doctrine/ORM/EntityManager.php [hunk 8]
```diff
@@ -139,7 +139,7 @@
      * @param \Doctrine\ORM\Configuration   $config
      * @param \Doctrine\Common\EventManager $eventManager
      */
-    protected function __construct(Connection $conn, Configuration $config, EventManager $eventManager)
+    public function __construct(Connection $conn, Configuration $config, EventManager $eventManager)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/EntityManager.php]: @usmanzafar was this made `public` on purpose?

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by usmanzafar** [comment]
Hey @Ocramius having that method to public was a major mistake at my end. I have updated the mock to create an EM via static method. 
Secondly, after this change to mock i was getting an error/exception of  "Cannot use different EventManager instances for EntityManager and Connection" - i have fixed that issue and updated the creation of EntityManager via Mock so that eventManager and connection's event manager remain similar where applicable. If you believe that this is outside the scope of the patch then let me know i am happy to take them away from tests and keep EntityManager / EntityManagerMock changes.
Thanks

**Review by Ocramius — APPROVED** [review]
LGTM :+1:

**Comment by Ocramius** [comment]
> Lastly, as per the new changes merged into develop the test `OnClassMetadataNotFoundEventArgsTest::testEventArgsMutability` is creating a mock of a final class which generates a warning in phpunit - shall we ignore this or do i need to update this - let me know your thoughts ?

@usmanzafar is it related to this patch?

**Comment by usmanzafar** [comment]
@Ocramius   Technically no, visually I was being annoyed looking at the fact that final class is being mocked  haha. I think we are good man. so happy to be part of this project. Happy to see if this gets merged to develop :)

**Comment by Ocramius** [comment]
:ship: then!

Thanks @usmanzafar \o/

**Pull request merged**